server/controllers/users.js

```diff
@@ -106,11 +108,25 @@ export const twoFactorAuthAndUpdateToken = async (req, res, next) => {
     return;
   }
 
-  // we need to verify the 2FA code before returning the token
-  const verified = verifyTwoFactorAuthenticatorCode(user.twoFactorAuthToken, twoFactorAuthenticatorCode);
-  if (!verified) {
-    return next(new Unauthorized('Two-factor authentication code failed. Please try again'));
+  if (twoFactorAuthenticatorCode) {
+    // if there is a 2FA code, we need to verify it before returning the token
+    const verified = verifyTwoFactorAuthenticatorCode(user.twoFactorAuthToken, twoFactorAuthenticatorCode);
+    if (!verified) {
+      return next(new Unauthorized('Two-factor authentication code failed. Please try again'));
+    }
+  } else if (recoveryCode) {
+    // or if there is a recovery code try to verify it
+    const { twoFactorAuthRecoveryCodes } = user;
```

Comment:
In `verifyTwoFactorAuthenticatorCode`, you're not extracting the value from the user object. You should stay consistent in your style and use `user.twoFactorAuthRecoveryCodes` directly.